sky/backends/backend_utils.py

```diff
@@ -2415,60 +2424,144 @@ def _refresh_cluster(cluster_name):
+        replica_info: List[Dict[str, Any]]) -> status_lib.ServiceStatus:
+    status2num = collections.Counter([i['status'] for i in replica_info])
+    # If one replica is READY, the service is READY.
+    if status2num[status_lib.ReplicaStatus.READY] > 0:
+        return status_lib.ServiceStatus.READY
+    if (status2num[status_lib.ReplicaStatus.FAILED] +
+            status2num[status_lib.ReplicaStatus.FAILED_CLEANUP] > 0):
+        return status_lib.ServiceStatus.FAILED
+    return status_lib.ServiceStatus.REPLICA_INIT
+
+
+def _check_controller_status_and_set_service_status(
+        service_name: str, cluster_name: str) -> Optional[str]:
+    cluster_record = global_user_state.get_cluster_from_name(cluster_name)
+    if (cluster_record is None or
+            cluster_record['status'] != status_lib.ClusterStatus.UP):
+        global_user_state.set_service_status(
+            service_name, status_lib.ServiceStatus.CONTRLLER_FAILED)
+        return f'Controller cluster {cluster_name!r} is not exist or UP.'
+    return None
+
+
+def _refresh_service_record_no_lock(
+        service_name: str) -> Tuple[Optional[Dict[str, Any]], Optional[str]]:
+    record = global_user_state.get_service_from_name(service_name)
+    if record is None:
+        return None, None
+
+    try:
+        check_network_connection()
+    except exceptions.NetworkError:
+        return record, 'Failed to refresh replica info due to network error.'
+
+    if not record['endpoint']:
+        # Service controller is still initializing. Skipped refresh status.
+        return record, None
+
+    controller_cluster_name = record['controller_cluster_name']
+    handle = global_user_state.get_handle_from_cluster_name(
+        controller_cluster_name)
+    assert handle is not None
+    backend = get_backend_from_handle(handle)
+    assert isinstance(backend, backends.CloudVmRayBackend)
+
+    code = serve_lib.ServeCodeGen.get_replica_info()
+    returncode, replica_info_payload, stderr = backend.run_on_head(
+        handle,
+        code,
+        require_outputs=True,
+        stream_logs=False,
+        separate_stderr=True)
+    if returncode != 0:
+        msg = _check_controller_status_and_set_service_status(
+            record['name'], controller_cluster_name)
+        if msg is None:
+            msg = ('Failed to refresh replica info from the controller. '
+                   f'Using the cached record. Reason: {stderr}')
+        return record, msg
+
+    replica_info = serve_lib.load_replica_info(replica_info_payload)
+    record['replica_info'] = replica_info
+
+    msg = None
+    if record['status'] != status_lib.ServiceStatus.SHUTTING_DOWN:
+        new_status = _service_status_from_replica_info(replica_info)
+        if (record['uptime'] < 0 and
+                record['status'] != status_lib.ServiceStatus.READY and
+                new_status == status_lib.ServiceStatus.READY):
+            code = serve_lib.ServeCodeGen.get_uptime()
```

Comment:
Is it possible to merge this code with the above 
```
code = serve_lib.ServeCodeGen.get_replica_info()
```
The reason is each ssh remote execution causes some latency. We probably don't want to slow down `sky serve status` too much.